lib/keypair.rb

```diff
@@ -66,11 +84,26 @@ def self.current
   #
   # @see https://www.imsglobal.org/spec/security/v1p0/#h_key-set-url
   def self.keyset
+    valid_keys = valid.order(expires_at: :asc).to_a
+    # If we don't have any keys or if we don't have a future key (i.e. the last key is the current key)
+    if valid_keys.last.nil? || valid_keys.last.not_before <= Time.zone.now
+      # There is an automatic fallback to Time.zone.now if not_before is not set
+      valid_keys << create!(not_before: valid_keys.last&.not_after)
+    end
+
     {
-      keys: valid.order(created_at: :desc).map(&:public_jwk_export)
+      keys: valid_keys.map(&:public_jwk_export)
     }
   end
 
+  # @return [Hash] a cached version of the keyset
+  # @see #keyset
+  def self.cached_keyset
+    Rails.cache.fetch('keypairs/Keypair/keyset', expires_in: 12.hours) do
```

Comment:
I think that many of our apps don't have a proper rails cache setup. Probably it's not so bad since that it will just have to make the SQL query again. But maybe it can help to be able to configure what cache to use. So something like `Keypairs.cache_store = Rails.cache`? But maybe that's for a next version again where we can maybe also set the rotation interval in a config file.

Reply:
I think this should indeed be something in the future. It also wouldn't really help, since it's probably less work to set up the Rails cache properly than creating a completely new cache just for the keypairs. If we actually want the caching in production, we should just make sure the Rails cache is setup properly, not create a completely new cache just for this.